sphinx/source/docs/user_guide/server.rst

```diff
@@ -347,6 +336,86 @@ In this case you might have code similar to:
 And similar code to load the JavaScript implementation for a custom model
 from ``models/custom.js``
 
+.. _userguide_server_template:
+
+Customising the Application's Jinja Template
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As described above in :ref:`userguide_server_applications_directory`, you can override
+the default Jinja template used by the Bokeh server to generate the HTML code served to
+the user's browser.
+
+This opens up the possibility of managing the layout of the application in the client's
+browser using CSS, as well as making use of other Javascript libraries alongside BokehJS.
+
+See the `Jinja Project Documentation`_ for more details on how Jinja templating works.
+
+Embedding Figures in the Template
+'''''''''''''''''''''''''''''''''
+
+In the main thread of the Bokeh application, i.e. ``main.py``, any Bokeh figures
+that are going to be referenced in the templated code need to have their ``name``
+attribute set, and be added to the current document root.
+
+.. code-block:: python
+
+    from bokeh.plotting import curdoc
+
+    # templates can refer to a configured name value
+    plot = figure(name="bokeh_jinja_figure")
+
+    curdoc().add_root(plot)
+
+Then, in the corresponding Jinja template code, the figure may be referenced via the
+``roots`` template parameter, using the figure's ``name``, i.e.
+
+.. code-block:: html
+
+    {% extends base %}
+
+    {% block contents %}
+    <div>
+        {{ embed(roots.bokeh_jinja_figure) }}
+    </div>
+    {% endblock %}
+
+Defining Custom Variables
+'''''''''''''''''''''''''
+
+Custom variables can be passed to the template via the ``curdoc().template_variables``
+dictionary in place:
+
+.. code-block:: python
+
+    # set a new single key/value
+    curdoc().template_variables["user_id"] = user_id
+
+    # or update multiple at once
+    curdoc().template_variables.update(first_name="Mary", last_name="Jones")
+
+Custom Jinja templates and Multiple Server Threads
+''''''''''''''''''''''''''''''''''''''''''''''''''
+Due to the nature of Bokeh's server-client communication, it is possible that a client
+will connect to a server thread that doesn't properly identify the Document's figure(s)
+to the client. This can only occur when there are multiple server threads (i.e.
+``--num-procs`` > 1), or there are multiple server instances behind a load balancer
+without a session persistence or stickiness mechanism.
```

Comment:
We should try to be precise. These are different *processes*, not different threads within a single process.  Also the actual issue is that, when there are multiple processes, the original HTTP connection to one server process generates a second, websocket connection, which may land on a completely different process. However while there are definitely things that this affects, e.g. not having access to the HTTP request, I am surprised that a template does not work. I don't think I was aware of that. Can we table this section for now, and open a separate issue to investigate this? It may be that there is a solution to be had, rather than adding a warning (that necessarily has to go fairly deep in the weeds).

Reply:
Sure, I'm not terribly familiar with the underlying threading model of Tornado, so the explanation of the process <-> websocket mismatch sounds as plausible to me as anything else ;)

I'm happy to drop this section for now, but I suggest we carry on discussing the issue in #9435, as it has the context from my previous investigation.